vertx-web/src/main/java/io/vertx/ext/web/handler/impl/StaticHandlerImpl.java

```diff
@@ -794,5 +753,139 @@ boolean isOutOfDate() {
       return System.currentTimeMillis() - createDate > cacheEntryTimeout;
     }
 
+    public boolean isMissing() {
+      return  props == null;
+    }
+  }
+
+  private static class FSTune {
+    // These members are all related to auto tuning of synchronous vs asynchronous file system access
+    private static int NUM_SERVES_TUNING_FS_ACCESS = 1000;
+
+    // these variables are read often and should always represent the
+    // real value, no caching should be allowed
+    private volatile boolean enabled = DEFAULT_ENABLE_FS_TUNING;
+    private volatile boolean useAsyncFS;
+
+    private long totalTime;
+    private long numServesBlocking;
+    private long nextAvgCheck = NUM_SERVES_TUNING_FS_ACCESS;
+    private long maxAvgServeTimeNanoSeconds = DEFAULT_MAX_AVG_SERVE_TIME_NS;
+    private boolean alwaysAsyncFS = DEFAULT_ALWAYS_ASYNC_FS;
+
+    boolean enabled() {
+      return enabled;
+    }
+
+    boolean useAsyncFS() {
+      return alwaysAsyncFS || useAsyncFS;
+    }
+
+    synchronized void setEnabled(boolean enabled) {
+      this.enabled = enabled;
+      if (!enabled) {
+        reset();
+      }
+    }
+
+    void setAlwaysAsyncFS(boolean alwaysAsyncFS) {
+      this.alwaysAsyncFS = alwaysAsyncFS;
+    }
+
+    synchronized void update(long start, long end) {
+      long dur = end - start;
+      totalTime += dur;
+      numServesBlocking++;
+      if (numServesBlocking == Long.MAX_VALUE) {
+        // Unlikely.. but...
+        reset();
+      } else if (numServesBlocking == nextAvgCheck) {
+        double avg = (double) totalTime / numServesBlocking;
+        if (avg > maxAvgServeTimeNanoSeconds) {
+          useAsyncFS = true;
+          log.info("Switching to async file system access in static file server as fs access is slow! (Average access time of " + avg + " ns)");
+          enabled = false;
+        }
+        nextAvgCheck += NUM_SERVES_TUNING_FS_ACCESS;
+      }
+    }
+
+    synchronized void reset() {
+      nextAvgCheck = NUM_SERVES_TUNING_FS_ACCESS;
+      totalTime = 0;
+      numServesBlocking = 0;
+    }
+  }
+
+  private class FSPropsCache {
+    private Map<String, CacheEntry> propsCache;
```

Comment:
I am assuming that FSPropsCache is accessed concurrently? there does not appear to be any memory barriers to accessing the underlying cache data structure (java.util.LinkedHashMap). I looks like access to the cache is racy, unless there is some synchronization outside the scope of this cache impl?

Reply:
no, it's only accessed from the event loop, like it was previously but now all cache operations are encapsulated on a internal class

Reply:
If an instance of StaticHandlerImpl is not concurrently accessed (and by extension FSPropsCache), how do we explain the ~100x speedup observed? Previously, there were synchronized methods, which aquire intrinsic lock for that method's object. If there is no contention on that object's monitor (e.g. by each thread having it's own  StaticHandlerImpl instance ) then aquire/release of instrinic locks should be fast.

Reply:
The issue wasn't the cache per se. The cache is correct (it was exactly the same as before but now it's wrapped on a object to keep the code more readable). The locking issue happens at the classpath file resolving algorithm. Now before if a file was missing, we didn't add it to the LRU cache as we should, which would mean that on every missing file (with cache active) we would hit the lock and filesystem all the time.

Now the fix was, to not throw away the misses, by doing this, we can avoid going to the file system and therefore avoid the resolver.

Reply:
I see, thank you for the explanation